operator/pkg/model/ingestion/gamma.go

```diff
@@ -0,0 +1,158 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package ingestion
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	gatewayv1 "sigs.k8s.io/gateway-api/apis/v1"
+	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+	"sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+
+	"github.com/cilium/cilium/operator/pkg/gateway-api/helpers"
+	"github.com/cilium/cilium/operator/pkg/model"
+)
+
+// Input is the input for GatewayAPI.
+type GammaInput struct {
+	HTTPRoutes      []gatewayv1.HTTPRoute
+	ReferenceGrants []gatewayv1beta1.ReferenceGrant
+	Services        []corev1.Service
+}
+
+// GammaHTTPRoutes takes a GammaInput and gives back the associated HTTP Listeners
+// It does not support TLS Routes because GAMMA is only for cleartext config -
+// it is assumed that any TLS will be performed transparently by the underlying
+// implementation in the spec.
+func GammaHTTPRoutes(input GammaInput) []model.HTTPListener {
+	// GAMMA processing process:
+	// Process HTTPRoutes
+	var resHTTP []model.HTTPListener
+
+	// Search algorithm:
+	// Loop through all HTTPRoutes in input, validate that each is parented by a
+	//   svc that's also in input, in the same namespace, etc.
+	// Add each parentRef service to some tracker to know if we need to create
+	//   a new HTTPListener or not.
+	// Add the rules from the HTTPRoute to the relevant HTTPListener.
+	//   unresolved: How to order these rules?
+	// ReferenceGrants are only relevant for backends, I think, because parents
+	//  _can_ be across namespace boundaries, but that makes them a consumer
+	// route, not a producer one.
+
+	// Set of services that will be parents for these HTTPRoutes
+	parentServices := make(map[types.NamespacedName]model.FullyQualifiedResource)
+
+	for _, hr := range input.HTTPRoutes {
+		// First, we find which parentRefs are Services, so that we can add the
+		// route rules to Listeners for those Services.
+		var gammaParents []gatewayv1.ParentReference
+		for _, parent := range hr.Spec.ParentRefs {
+			if helpers.IsGammaService(parent) {
+				gammaParents = append(gammaParents, parent)
+			}
+		}
+
+		// We shouldn't be able to do this, because the reconciliation should
+		// screen out HTTPRoutes with zero GAMMA parents.
+		if len(gammaParents) == 0 {
+			continue
+		}
+
+		for _, gp := range gammaParents {
+
+			if gp.Name == "" {
+				continue
+			}
+
+			parentName := types.NamespacedName{
+				Name: string(gp.Name),
+			}
+
+			if gp.Namespace != nil {
+				parentName.Namespace = string(*gp.Namespace)
+			}
+
+			parentSvc, err := getMatchingService(parentName.Name, parentName.Namespace, hr.GetNamespace(), input.Services)
+			if err != nil {
+				fmt.Printf("Can't find parent Service %s/%s in input\n", parentName.Namespace, parentName.Name)
```

Comment:
fmt.Printf? :thinking:  Did you mean `log`?

Reply:
🤦 One day I'll remember to remove all the debugging prints before submitting the PR.